docs/configuration.rst
.. _configuration_customisation:

Configuration and customisation
===============================

Tutor offers plenty of possibilities for platform customisation out of the box. There are two main ways in which the base Open edX installation can be customised:

a. Modifying the Tutor :ref:`configuration parameters <configuration>`.
b. Modifying the :ref:`Open edX docker image <customise>` that runs the Open edX platform.

This section does not cover :ref:`plugin development <plugins>`. For simple changes, such as modifying the ``*.env.yml`` files or the edx-platform settings, *you should not fork edx-platform or tutor*! Instead, you should create a simple :ref:`plugin for Tutor <plugins_yaml>`.

.. _configuration:

Configuration
-------------

With Tutor, all Open edX deployment parameters are stored in a single ``config.yml`` file. This is the file that is generated when you run ``tutor local launch`` or ``tutor config save``. To view the content of this file, run::

    cat "$(tutor config printroot)/config.yml"

By default, this file contains only the required configuration parameters for running the platform. Optional configuration parameters may also be specified to modify the default behaviour. To do so, you can edit the ``config.yml`` file manually::

    vim "$(tutor config printroot)/config.yml"

Alternatively, you can set each parameter from the command line::

    tutor config save --set PARAM1=VALUE1 --set PARAM2=VALUE2

Or from the system environment::

    export TUTOR_PARAM1=VALUE1

Once the base configuration is created or updated, the environment is automatically re-generated. The environment is the set of all files required to manage an Open edX platform: Dockerfile, ``lms.env.yml``, settings files, etc. You can view the environment files in the ``env`` folder::

    ls "$(tutor config printroot)/env"

With an up-to-date environment, Tutor is ready to launch an Open edX platform and perform usual operations. Below, we document some of the configuration parameters.

Individual service activation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- ``RUN_MEILISEARCH`` (default: ``true``)
- ``RUN_MONGODB`` (default: ``true``)
- ``RUN_MYSQL`` (default: ``true``)
- ``RUN_REDIS`` (default: ``true``)
- ``RUN_SMTP`` (default: ``true``)
- ``ENABLE_HTTPS`` (default: ``false``)

Every single Open edX service may be (de)activated at will by these configuration parameters. This is useful if you want, for instance, to distribute the various Open edX services on different servers.

Docker
~~~~~~

.. _docker_images:

Custom images
*************

- ``DOCKER_IMAGE_OPENEDX`` (default: ``"{{ DOCKER_REGISTRY }}overhangio/openedx:{{ TUTOR_VERSION }}"``)

This configuration parameter defines the name of the Docker image to run for the lms and cms containers. By default, the Docker image tag matches the Tutor version it was built with.

- ``DOCKER_IMAGE_OPENEDX_DEV`` (default: ``"openedx-dev:{{ TUTOR_VERSION }}"``)

This configuration parameter defines the name of the Docker image to run the development version of the lms and cms containers.  By default, the Docker image tag matches the Tutor version it was built with.

.. https://hub.docker.com/r/devture/exim-relay/tags

- ``DOCKER_IMAGE_CADDY`` (default: ``"docker.io/caddy:2.6.2"``)

This configuration parameter defines which Caddy Docker image to use.

- ``DOCKER_IMAGE_MEILISEARCH`` (default: ``"docker.io/getmeili/meilisearch:v1.8.4"``)

This configuration parameter defines which Meilisearch Docker image to use.

- ``DOCKER_IMAGE_MONGODB`` (default: ``"docker.io/mongo:7.0.7"``)

This configuration parameter defines which MongoDB Docker image to use.

.. https://hub.docker.com/_/mysql/tags?page=1&name=8.0

- ``DOCKER_IMAGE_MYSQL`` (default: ``"docker.io/mysql:8.4.0"``)

This configuration parameter defines which MySQL Docker image to use.

.. https://hub.docker.com/_/redis/tags

- ``DOCKER_IMAGE_REDIS`` (default: ``"docker.io/redis:7.4.5"``)

This configuration parameter defines which Redis Docker image to use.

.. https://hub.docker.com/r/devture/exim-relay/tags

- ``DOCKER_IMAGE_SMTP`` (default: ``"docker.io/devture/exim-relay:4.96-r1-0``)

This configuration parameter defines which Simple Mail Transfer Protocol (SMTP) Docker image to use.

- ``DOCKER_IMAGE_PERMISSIONS`` (default: ``"{{ DOCKER_REGISTRY }}overhangio/openedx-permissions:{{ TUTOR_VERSION }}"``)

This configuration parameter defines the Docker image to be used for setting file permissions. The default image sets all containers to be run as unprivileged users.

Custom registry
***************

- ``DOCKER_REGISTRY`` (default: ``"docker.io/"``)

You may want to pull/push images from/to a custom docker registry. For instance, for a registry running on ``localhost:5000``, define::

    DOCKER_REGISTRY: localhost:5000/

(the trailing ``/`` is important)

.. _openedx_configuration:

Compose
*******

- ``DEV_PROJECT_NAME`` (default: ``"{{ TUTOR_APP }}_dev"``)

This configuration parameter sets the Development version of the Docker Compose project name.

- ``LOCAL_PROJECT_NAME`` (default: ``"{{ TUTOR_APP }}_local"``)

This configuration parameter sets the Local version of the Docker Compose project name.

Open edX customisation
~~~~~~~~~~~~~~~~~~~~~~

- ``EDX_PLATFORM_REPOSITORY`` (default: ``"https://github.com/openedx/edx-platform.git"``)

This defines the git repository from which you install Open edX platform code. If you run an Open edX fork with custom patches, set this to your own git repository. You may also override this configuration parameter at build time, by providing a ``--build-arg`` option.

<<<<<<< HEAD
- ``OPENEDX_COMMON_VERSION`` (default: ``"release/teak.3"``, or ``master`` in :ref:`Tutor Main <main>`)
=======
- ``OPENEDX_COMMON_VERSION`` (default: ``"release/teak.2"``, or ``master`` in :ref:`Tutor Main <main>`)
>>>>>>> origin/refactor/config-system-refactor

This defines the default version that will be pulled from all Open edX git repositories.

- ``EDX_PLATFORM_VERSION`` (default: the value of ``OPENEDX_COMMON_VERSION``)

This defines the version that will be pulled from just the Open edX platform git repositories. You may also override this configuration parameter at build time, by providing a ``--build-arg`` option.

- ``OPENEDX_CMS_UWSGI_WORKERS`` (default: ``2``)
- ``OPENEDX_LMS_UWSGI_WORKERS`` (default: ``2``)

By default, there are 2 `uwsgi worker processes <https://uwsgi-docs.readthedocs.io/en/latest/Options.html#processes>`__ to serve requests for the LMS and the CMS. However, each worker requires upwards of 500 Mb of RAM. You should reduce this value to 1 if your computer/server does not have enough memory.

- ``OPENEDX_CELERY_REDIS_DB`` (default: ``0``)
- ``OPENEDX_CACHE_REDIS_DB`` (default: ``1``)

These two configuration parameters define which Redis database to use for Open edX cache and celery task.

.. _openedx_extra_pip_requirements:

- ``OPENEDX_EXTRA_PIP_REQUIREMENTS`` (default: ``[]``)

Define extra pip packages that are going to be installed for edx-platform.

- ``NPM_REGISTRY`` (default: ``"https://registry.npmjs.org/"``)

This defines the registry from which you'll be pulling NPM packages when building Docker images. Like ``EDX_PLATFORM_REPOSITORY``, this can be overridden at build time with a ``--build-arg`` option.

- ``OPENEDX_AWS_ACCESS_KEY`` (default: ``""``)

This configuration parameter sets the Django setting ``AWS_ACCESS_KEY_ID`` in edx-platform's LMS, CMS, envs, and production.py for use by the library django-storages with Amazon S3.

- ``OPENEDX_AWS_SECRET_ACCESS_KEY`` (default: ``""``)

This configuration parameter sets the Django setting ``AWS_SECRET_ACCESS_KEY`` in edx-platform's LMS, CMS, envs, and production.py for use by the library django-storages with Amazon S3.

- ``OPENEDX_MYSQL_DATABASE`` (default: ``"openedx"``)

This configuration parameter sets the name of the MySQL Database to be used by the Open edX Instance.

- ``OPENEDX_MYSQL_USERNAME`` (default: ``"openedx"``)

This configuration parameter sets the username associated with the MySQL Database.

CMS OAUTH2 SSO
~~~~~~~~~~~~~~

- ``CMS_OAUTH2_KEY_SSO`` (default: ``"cms-sso"``)

This defines the Studio's (CMS) OAUTH 2.0 Login (Key or Client ID) for SSO in the production environment.

- ``CMS_OAUTH2_KEY_SSO_DEV`` (default: ``"cms-sso-dev"``)

This defines the Studio's (CMS) OAUTH 2.0 Login (Key or Client ID) for SSO in the development environment.

For more information, see `Enabling OAuth for Studio login <https://github.com/openedx/edx-platform/blob/master/docs/guides/studio_oauth.rst>`__.

JWTs
~~~~

- ``JWT_COMMON_AUDIENCE`` (default: ``"openedx"``)
- ``JWT_COMMON_ISSUER`` (default: ``"{% if ENABLE_HTTPS %}https{% else %}http{% endif %}://{{ LMS_HOST }}/oauth2"``)
- ``JWT_COMMON_SECRET_KEY`` (default: ``"{{ OPENEDX_SECRET_KEY }}"``)

These configuration parameters are rendered into the ``JWT_AUTH`` dictionary with keys ``JWT_AUDIENCE``, ``JWT_ISSUER``, and ``JWT_SECRET_KEY``, respectively. These parameters may be changed in order to create a custom user login for testing purposes.

Vendor services
~~~~~~~~~~~~~~~

Caddy
*****

- ``CADDY_HTTP_PORT`` (default: ``80``)
- ``ENABLE_WEB_PROXY`` (default: ``true``)

`Caddy <https://caddyserver.com>`__ is a web server used in Tutor both as a web proxy and for the generation of SSL/TLS certificates at runtime. Port indicated by ``CADDY_HTTP_PORT`` is exposed on the host, in addition to port 443. If ``ENABLE_WEB_PROXY`` is set to ``false`` then we assume that SSL termination does not occur in the Caddy container and only ``CADDY_HTTP_PORT`` is exposed on the host.

MySQL
*****

- ``RUN_MYSQL`` (default: ``true``)
- ``MYSQL_HOST`` (default: ``"mysql"``)
- ``MYSQL_PORT`` (default: ``3306``)
- ``MYSQL_ROOT_USERNAME`` (default: ``"root"``)
- ``MYSQL_ROOT_PASSWORD`` (default: randomly generated) Note that you are responsible for creating the root user if you are using a managed database.

By default, a running Open edX platform deployed with Tutor includes all necessary 3rd-party services, such as MySQL, MongoDb, etc. But it's also possible to store data on a separate database, such as `Amazon RDS <https://aws.amazon.com/rds/>`_. For instance, to store data on an external MySQL database set the following configuration::

    RUN_MYSQL: false
    MYSQL_HOST: yourhost
    MYSQL_ROOT_USERNAME: <root user name>
    MYSQL_ROOT_PASSWORD: <root user password>

.. note::
    When configuring an external MySQL database, please make sure it is using version 8.4.

Meilisearch
***********

- ``MEILISEARCH_URL`` (default: ``"http://meilisearch:7700"``): internal URL used for backend-to-backend communication.

CHANGELOG.md
# Changelog

This file includes a history of past releases. Changes that were not yet added to a release are in the [changelog.d/](./changelog.d) folder.

<!--
⚠️ DO NOT ADD YOUR CHANGES TO THIS FILE! (unless you want to modify existing changelog entries in this file)
Changelog entries are managed by scriv. After you have made some changes to Tutor, create a changelog entry with:

    make changelog-entry

Edit and commit the newly-created file in changelog.d.

If you need to create a new release, create a separate commit just for that. It is important to respect these
instructions, because git commits are used to generate release notes:
  - Collect changelog entries with `make changelog`
  - The title of the commit should be the same as the CHANGELOG.md file section title: "vX.Y.Z (year-month-day)".
  - The commit message should be copy-pasted from the release section.
  - Have a look at other release commits for reference.
-->

<!-- scriv-insert-here -->

<<<<<<< HEAD
<a id='changelog-20.0.4'></a>
## v20.0.4 (2025-12-22)

- [Chore] Update hatch definition for version. (by @mlabeeb03)

- [Improvement] Add ability to not delete namespace while running `tutor k8s delete` command. (by @mlabeeb03)

- [Security] Add upstream security fix as patch in Open edX image (by @ahmed-arb)

[Security] Bump urllib3 from 2.5.0 to 2.6.0 to address Dependabot alerts.

<a id='changelog-20.0.3'></a>
## v20.0.3 (2025-12-03)

- [Bugfix] Fix packaging issue where a stray `venv/` directory was included in the source distribution (sdist)
for previous releases. The build configuration now correctly excludes the `venv/` folder from package archives. (by @ahmed-arb)

- [Feature] Add tutor livedeps to the list of plugins available by default. (by @mlabeeb03)

- [Feature] Update OPENEDX_COMMON_VERSION to teak.3 tag (by @ahmed-arb)

=======
>>>>>>> origin/refactor/config-system-refactor
<a id='changelog-20.0.2'></a>
## v20.0.2 (2025-10-20)

 - [Feature] Add support for pruning unused Kubernetes ConfigMaps with `--prune-configmaps` via the Tutor CLI. (by @dkaliberda)

- [Security] Switch to a cryptographically secure number generator for the `random_string` function. (by @samuelallan72)

- [Improvement] Upgrade node to v20.19.5 in Open edX Docker image. (by @dawoudsheraz)

- [Improvement] Migrate from pylint and black to ruff. (by @ahmed-arb)

<a id='changelog-20.0.1'></a>
## v20.0.1 (2025-08-22)

- [Feature] Add max-log-requests parameter for command `tutor k8s logs`. (by @mlabeeb03)

- [Improvement] Redis version upgraded to 7.4.5. (by @CodeWithEmad)

- [Feature] Update OPENEDX_COMMON_VERSION to teak.2 tag (by @ahmed-arb)

<a id='changelog-20.0.0'></a>
## v20.0.0 (2025-06-16)

- 💥[Depreciation] Remove the unnecessary site-configuration script in favour of the [`create_or_update_site_configuration`](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/site_configuration/management/commands/create_or_update_site_configuration.py) management command provided in edx-platform. (by @Danyal-Faheem)

- 💥[Feature] Upgrade to Teak. (by @ahmed-arb)

<a id='changelog-19.0.5'></a>
## v19.0.5 (2025-06-10)

- [Feature] Enable the "contrib" plugin index by default. This enables platform users to view 3rd party plugins without having to run `tutor plugins index add contrib`. (by @regisb)
- [Feature] Automatically update the plugin index cache on all commands that need it. Including: `plugins install/upgrade/search`. This makes it slightly easier for end users to use plugins. (by @regisb)

<a id='changelog-19.0.4'></a>
## v19.0.4 (2025-05-27)

- [Improvement] When running `local/dev start/run`, run `dev/local stop` less frequently. Also, make sure that we detect only running docker compose projects associated to Tutor. (by @regisb)
- [Improvement] Do not wait for non-tutor jobs while running k8s. (by @mlabeeb03)
- [Bugfix] Support plugin installation with unexpected `pip` path and/or uv. (by @regisb)

<a id='changelog-19.0.3'></a>
## v19.0.3 (2025-05-19)

- [Improvement] Migrate packaging from setup.py/setuptools to pyproject.toml/hatch. This should not be a breaking change for most users. (by @regisb)

- [Improvement] Add hatch_build.py in sdist target to fix the installation issues (by @dawoudsheraz)

- [Improvement] MEILISEARCH_HOST enables Tutor to configure meilisearch host
and expose it to be used by other plugins. For example MEILISEARCH_HOST can be
optionally used to configure ingress controller in openedx-k8s-harmony like
LMS_HOST etc.

- [Bugfix] Update lms and cms dev jobs to use correct django settings (by @dawoudsheraz)

- [Improvement] Add support for `docker-compose.prod.override.yml` when using
  `tutor local` commands

- [Feature] Add tutor deck to the list of plugins installed by default. (by @mlabeeb03)

- [Improvement] Stop `dev` platform on `local run` commands. Trigger `COMPOSE_PROJECT_STARTED` only once per tutor runtime. (by @mlabeeb03)

 - [Feature] Add new command `tutor dev hosts` that displays status of services. (by @mlabeeb03)

- [Feature] Update OPENEDX_COMMON_VERSION to sumac.3 tag (by @dawoudsheraz)

<a id='changelog-19.0.2'></a>
## v19.0.2 (2025-02-12)


tutor/env.py
from __future__ import annotations

import os
import re
import shutil
import typing as t
from copy import deepcopy

import importlib_resources
import jinja2

from tutor import exceptions, fmt, hooks, plugins, utils
from tutor.__about__ import __app__, __version__, __version_suffix__
from tutor.edops import modules as edops_modules
from tutor.types import Config, ConfigValue

TEMPLATES_ROOT = str(importlib_resources.files("tutor") / "templates")
VERSION_FILENAME = "version"
BIN_FILE_EXTENSIONS = [
    ".ico",
    ".jpg",
    ".otf",
    ".patch",
    ".png",
    ".ttf",
    ".webp",
    ".woff",
    ".woff2",
]
JinjaFilter = t.Callable[..., t.Any]


def _prepare_environment() -> None:
    """
    Prepare environment by adding core data to filters.
    """
    # Core template targets
    hooks.Filters.ENV_TEMPLATE_TARGETS.add_items(
        [
            ("apps/", ""),
            ("build/", ""),
            ("dev/", ""),
            ("k8s/", ""),
            ("local/", ""),
            (VERSION_FILENAME, ""),
            ("kustomization.yml", ""),
        ],
    )
    # Template filters
    hooks.Filters.ENV_TEMPLATE_FILTERS.add_items(
        [
            ("common_domain", utils.common_domain),
            ("encrypt", utils.encrypt),
            ("list_if", utils.list_if),
            ("long_to_base64", utils.long_to_base64),
            ("random_string", utils.random_string),
            ("reverse_host", utils.reverse_host),
            ("rsa_import_key", utils.rsa_import_key),
            ("rsa_private_key", utils.rsa_private_key),
            ("uuid", utils.uuid),
            ("uid_master_hash", utils.uid_master_hash),
        ],
    )
    # Template variables
    hooks.Filters.ENV_TEMPLATE_VARIABLES.add_items(
        [
            ("HOST_USER_ID", utils.get_user_id()),
            ("TUTOR_APP", __app__.replace("-", "_")),
            ("TUTOR_VERSION", __version__),
            # We assume that TUTOR_VERSION is always MAJOR.MINOR.PATCH.
            # We assume that MAJOR and MINOR are both integers.
            # We make no assumptions about PATCH, and thus do not parse it for plugins to use.
            ("TUTOR_VERSION_MAJOR", int(__version__.split(".")[0])),
            ("TUTOR_VERSION_MINOR", int(__version__.split(".")[1])),
            ("TUTOR_BRANCH_IS_MAIN", __version_suffix__ == "main"),
        ],
    )


_prepare_environment()


class JinjaEnvironment(jinja2.Environment):
    loader: jinja2.FileSystemLoader

    def __init__(self) -> None:
        template_roots = hooks.Filters.ENV_TEMPLATE_ROOTS.apply([TEMPLATES_ROOT])
        loader = jinja2.FileSystemLoader(template_roots)
        super().__init__(loader=loader, undefined=jinja2.StrictUndefined)

    def read_str(self, template_name: str) -> str:
        return self.read_bytes(template_name).decode()

    def read_bytes(self, template_name: str) -> bytes:
        with open(self.find_os_path(template_name), "rb") as f:
            return f.read()

    def find_os_path(self, template_name: str) -> str:
        path = template_name.replace("/", os.sep)
        for templates_root in self.loader.searchpath:
            full_path = os.path.join(templates_root, path)
            if os.path.exists(full_path):
                return full_path
        raise ValueError("Template path does not exist")


class Renderer:
    def __init__(self, config: t.Optional[Config] = None):
        self.config = deepcopy(config or {})

        # Create environment with extra filters and globals
        self.environment = JinjaEnvironment()

        # Filters
        plugin_filters = hooks.Filters.ENV_TEMPLATE_FILTERS.iterate()
        for name, func in plugin_filters:
            if name in self.environment.filters:
                fmt.echo_alert(f"Found conflicting template filters named '{name}'")
            self.environment.filters[name] = func
        self.environment.filters["walk_templates"] = self.walk_templates

        # Globals
        for name, value in hooks.Filters.ENV_TEMPLATE_VARIABLES.iterate():
            if name in self.environment.globals:
                fmt.echo_alert(f"Found conflicting template variables named '{name}'")
            self.environment.globals[name] = value
        self.environment.globals["iter_values_named"] = self.iter_values_named
        self.environment.globals["patch"] = self.patch

    def iter_templates_in(self, *prefix: str) -> t.Iterable[str]:
        """
        The elements of `prefix` must contain only "/", and not os.sep.
        """
        full_prefix = "/".join(prefix)
        env_templates: list[str] = self.environment.loader.list_templates()
        for template in env_templates:
            if template.startswith(full_prefix):
                # Exclude templates that match certain patterns
                # Note that here we don't rely on the OS separator, as we are handling templates.
                if is_rendered(template):
                    yield template

    def iter_values_named(
        self,
        prefix: t.Optional[str] = None,
        suffix: t.Optional[str] = None,
        allow_empty: bool = False,
    ) -> t.Iterable[ConfigValue]:
        """
        Iterate on all config values for which the name match the given pattern.

        Note that here we only iterate on the values, not the key names. Empty
        values (those that evaluate to boolean `false`) will not be yielded, unless
        `allow_empty` is True.
        """
        for var_name, value in self.config.items():
            if prefix is not None and not var_name.startswith(prefix):
                continue
            if suffix is not None and not var_name.endswith(suffix):
                continue
            if not allow_empty and not value:
                continue
            yield value

    def walk_templates(self, subdir: str) -> t.Iterable[str]:
        """
        Iterate on the template files from `templates/<subdir>`.

        Yield:
            path: template path relative to the template root
        """
        yield from self.iter_templates_in(subdir)

    def patch(self, name: str, separator: str = "\n", suffix: str = "") -> str:
        """
        Render calls to {{ patch("...") }} in environment templates from plugin patches.
        """
        patches = []
        for patch in plugins.iter_patches(name):
            try:
                patches.append(self.render_str(patch))
            except exceptions.TutorError:
                fmt.echo_error(f"Error rendering patch '{name}':\n{patch}")
                raise
        rendered = separator.join(patches)
        if rendered:
            rendered += suffix
        return rendered

    def render_str(self, text: str) -> str:
        try:
            template = self.environment.from_string(text)
        except jinja2.exceptions.TemplateSyntaxError as e:
            raise exceptions.TutorError(f"Template syntax error: {e.args[0]}")
        return self.__render(template)

    def render_template(self, template_name: str) -> t.Union[str, bytes]:
        """
        Render a template file. Return the corresponding string. If it's a binary file
        (as indicated by its path), return bytes.

        The template_name *always* uses "/" separators, and is not os-dependent. Do not pass the result of
        os.path.join(...) to this function.
        """
        if not hooks.Filters.IS_FILE_RENDERED.apply(True, template_name):
            return self.environment.read_bytes(template_name)

        try:
            template = self.environment.get_template(template_name)
        except Exception:
            fmt.echo_error("Error loading template " + template_name)
            raise

        try:
            return self.__render(template)
        except (jinja2.exceptions.TemplateError, exceptions.TutorError):
            fmt.echo_error("Error rendering template " + template_name)
            raise
        except Exception:
            fmt.echo_error("Unknown error rendering template " + template_name)
            raise

    def render_all_to(self, dst: str, *prefix: str) -> None:
        """
        `prefix` can be used to limit the templates to render.
        """
        for template_name in self.iter_templates_in(*prefix):
            rendered = self.render_template(template_name)
            template_dst = os.path.join(dst, template_name.replace("/", os.sep))
            write_to(rendered, template_dst)

    def __render(self, template: jinja2.Template) -> str:
        try:
            return template.render(**self.config)
        except jinja2.exceptions.UndefinedError as e:
            raise exceptions.TutorError(f"Missing configuration value: {e.args[0]}")


class PatchRenderer(Renderer):
    """
    Render patches for print it.
    """

    def __init__(self, config: t.Optional[Config] = None):
        self.patches_locations: t.Dict[str, t.List[str]] = {}
        self.current_template: str = ""
        super().__init__(config)

    def render_template(self, template_name: str) -> t.Union[str, bytes]:
        """
        Set the current template and render template from Renderer.
        """
        self.current_template = template_name
        return super().render_template(self.current_template)

    def patch(self, name: str, separator: str = "\n", suffix: str = "") -> str:
        """
        Set the patches locations and render calls to {{ patch("...") }} from Renderer.
        """
        if not self.patches_locations.get(name):
            self.patches_locations.update({name: [self.current_template]})
        else:
            if self.current_template not in self.patches_locations[name]:
                self.patches_locations[name].append(self.current_template)

        # Store the template's name, and replace it with the name of this patch.
        # This handles the case where patches themselves include patches.
        original_template = self.current_template
        self.current_template = f"within patch: {name}"

        rendered_patch = super().patch(name, separator=separator, suffix=suffix)
        self.current_template = (
            original_template  # Restore the template's name from before.
        )
        return rendered_patch

    def render_all(self, *prefix: str) -> None:
        """
        Render all templates.
        """
        for template_name in self.iter_templates_in(*prefix):
            self.render_template(template_name)

    def print_patches_locations(self) -> None:
        """
        Print patches locations.
        """
        plugins_table: list[tuple[str, ...]] = [("PATCH", "LOCATIONS")]
        self.render_all()
        for patch, locations in sorted(self.patches_locations.items()):
            n_locations = 0
            for location in locations:
                if n_locations < 1:
                    plugins_table.append((patch, location))
                    n_locations += 1
                else:
                    plugins_table.append(("", location))

        fmt.echo(utils.format_table(plugins_table))


def is_rendered(path: str) -> bool:
    """
    Return whether the template should be rendered or not.

    If the path matches an include pattern, it is rendered. If not and it matches an
    ignore pattern, it is not rendered. By default, all files are rendered.
    """
    for include_pattern in hooks.Filters.ENV_PATTERNS_INCLUDE.iterate():
        if re.match(include_pattern, path):
            return True
    for ignore_pattern in hooks.Filters.ENV_PATTERNS_IGNORE.iterate():
        if re.match(ignore_pattern, path):
            return False
    return True


# Skip rendering some files that follow commonly-ignored patterns:
#
#   .*
#   *.pyc
#   __pycache__
#   partials
hooks.Filters.ENV_PATTERNS_IGNORE.add_items(
    [
        # Skip all hidden files
        r"(.*/)?\.",
        # Skip compiled python files
        r"(.*/)?__pycache__(/.*)?$",
        r".*\.pyc$",
        # Skip files from "partials" folders
        r"(.*/)?partials(/.*)?$",
        # 忽略所有 Open edX 相关模板（仅交付 zhjx 业务）
        r"apps/openedx(/.*)?$",
        r"build/openedx(/.*)?$",
    ]
)


def save(root: str, config: Config) -> None:
    """
    Save the full environment, including version information.
    """
    root_env = pathjoin(root)
    for src, dst in hooks.Filters.ENV_TEMPLATE_TARGETS.iterate():
        save_all_from(src, os.path.join(root_env, dst), config)
    edops_modules.render_modules(root_env, config)

    upgrade_obsolete(root)
    fmt.echo_info(f"Environment generated in {base_dir(root)}")


def upgrade_obsolete(_root: str) -> None:
    """
    Add here ad-hoc commands to upgrade the environment.
    """


def save_all_from(prefix: str, dst: str, config: Config) -> None:
    """
    Render the templates that start with `prefix` and store them with the same
    hierarchy at `dst`. Here, `prefix` can be the result of os.path.join(...).
    """
    renderer = Renderer(config)
    renderer.render_all_to(dst, prefix.replace(os.sep, "/"))


def write_to(content: t.Union[str, bytes], path: str) -> None:
    """
    Write some content to a path. Content can be either str or bytes.
    """
    utils.ensure_file_directory_exists(path)
    if isinstance(content, bytes):
        with open(path, mode="wb") as of_binary:
            of_binary.write(content)
    else:
        with open(path, mode="w", encoding="utf8", newline="\n") as of_text:
            of_text.write(content)


def render_file(config: Config, *path: str) -> t.Union[str, bytes]:
    """
    Return the rendered contents of a template.
    """
    renderer = Renderer(config)
    template_name = "/".join(path)
    return renderer.render_template(template_name)


def render_unknown(config: Config, value: t.Any) -> t.Any:
    """
    Render an unknown `value` object with the selected config.

    If `value` is a dict, its values are also rendered.
    """
    if isinstance(value, str):
        return render_str(config, value)
    if isinstance(value, dict):
        return {k: render_unknown(config, v) for k, v in value.items()}
    if isinstance(value, list):
        return [render_unknown(config, v) for v in value]
    return value


def render_str(config: Config, text: str) -> str:
    """
    Args:
        text (str)
        config (dict)

    Return:
        substituted (str)
    """
    return Renderer(config).render_str(text)


def check_is_up_to_date(root: str) -> None:
    if not is_up_to_date(root):
        fmt.echo_alert(
            f"The current environment stored at {base_dir(root)} is not up-to-date: it is at "
            f"v{current_version(root)} while the 'tutor' binary is at v{__version__}. You should upgrade "
            f"the environment by running:\n"
            f"\n"
            f"    tutor config save"
        )


def is_up_to_date(root: str) -> bool:
    """
    Check if the currently rendered version is equal to the current tutor version.
    """
    current = current_version(root)
    return current is None or current == __version__


def should_upgrade_from_release(root: str) -> t.Optional[str]:
    """
    Return the name of the currently installed release that we should upgrade from. Return None If we already run the
    latest release.
    """
    current = current_version(root)
    if current is None:
        return None
    current_as_int = int(current.split(".")[0])
    required_as_int = int(__version__.split(".", maxsplit=1)[0])
    if current_as_int >= required_as_int:
        return None
    return get_release(current)


def get_env_release(root: str) -> t.Optional[str]:
    """
    Return the Open edX release name from the current environment.

    If the current environment has no version, return None.
    """
    version = current_version(root)
    if version is None:
        return None
    return get_release(version)


def get_current_open_edx_release_name() -> str:
    """
    Return the release name associated to this package.
    """
    return get_release(__version__)


def get_release(version: str) -> str:
    return {
        "0": "ironwood",
        "3": "ironwood",
        "10": "juniper",
        "11": "koa",
        "12": "lilac",
        "13": "maple",
        "14": "nutmeg",
        "15": "olive",
        "16": "palm",
        "17": "quince",
        "18": "redwood",
        "19": "sumac",
        "20": "teak",
        "21": "ulmo",
    }[version.split(".", maxsplit=1)[0]]


def current_version(root: str) -> t.Optional[str]:
    """
    Return the current environment version. If the current environment has no version,
    return None.
    """
    path = pathjoin(root, VERSION_FILENAME)
    if not os.path.exists(path):
        return None
    with open(path, encoding="utf-8") as fi:
        return fi.read().strip()


def read_template_file(*path: str) -> str:
    """
    Read raw content of template located at `path`.

    The template may be located inside any of the template root folders.
    """
    return JinjaEnvironment().read_str("/".join(path))


def read_core_template_file(*path: str) -> str:
    """
    Read raw content of template located in tutor core template directory.
    """
    with open(os.path.join(TEMPLATES_ROOT, *path), encoding="utf-8") as fi:
        return fi.read()


def is_binary_file(path: str) -> bool:
    ext = os.path.splitext(path)[1]
    return ext in BIN_FILE_EXTENSIONS


def data_path(root: str, *path: str) -> str:
    """
    Return the file's absolute path inside the data directory.
    """
    return os.path.join(root_dir(root), "data", *path)


def pathjoin(root: str, *path: str) -> str:
    """
    Return the file's absolute path inside the environment.
    """
    return os.path.join(base_dir(root), *path)


def base_dir(root: str) -> str:
    """
    Return the environment base directory.
    """
    return os.path.join(root_dir(root), "env")


def root_dir(root: str) -> str:
    """
    Return the project root directory.
    """
    return os.path.abspath(root)


def delete_env_dir(root: str) -> None:
    env_path = base_dir(root)

    try:
        shutil.rmtree(env_path)
        fmt.echo_alert(f"Removed existing Tutor environment at: {env_path}")
    except PermissionError as e:
        raise exceptions.TutorError(
            f"Permission Denied while trying to remove existing Tutor environment at: {env_path}"
        ) from e
    except FileNotFoundError:
        fmt.echo_info(f"No existing Tutor environment to remove at: {env_path}")


@hooks.Actions.PLUGIN_UNLOADED.add()
def _delete_plugin_templates(plugin: str, root: str, _config: Config) -> None:
    """
    Delete plugin env files on unload.
    """
    targets = hooks.Filters.ENV_TEMPLATE_TARGETS.iterate_from_context(
        hooks.Contexts.app(plugin).name
    )
    for src, dst in targets:
        path = pathjoin(root, dst.replace("/", os.sep), src.replace("/", os.sep))
        if os.path.exists(path):
            fmt.echo_info(f"    env - removing folder: {path}")
            try:
                shutil.rmtree(path)
            except PermissionError as e:
                raise exceptions.TutorError(
                    f"Could not delete file {e.filename} from plugin {plugin} in folder {path}"
                )


@hooks.Filters.IS_FILE_RENDERED.add()
def _do_not_render_binary_files(result: bool, path: str) -> bool:
    """
    Determine whether a file should be rendered based on its binary nature.

    This function checks if a file is binary and updates the rendering decision accordingly.
    If the initial decision (`result`) is to render the file, but the file is detected as binary,
    the function will override this decision to prevent rendering.

    Parameters:
    - result (bool): The initial decision on whether the file should be rendered.
    - path (str): The file path to check for binary content.

    Returns:
    - bool: False if the file is binary and was initially set to be rendered, otherwise returns the initial decision.
    """
    if result and is_binary_file(path):
        result = False
    return result
